enable update after digest

aure-olli opened the pull request:
Hi

`hashlib` allows to make more updates after a digest, but this library does not allow that.
this PR allows it, by  making a copy of the MetroHash C++ object before calling `Finalize`.

In my specific case this is an optimization I'd like to use. In general, I think that this is more intuitive to not have to worry about that.

This costs an extra copy of the object (70 bytes for the 128 bits version if I counted well), which I think is extremely fast in C++.

aure-olli committed:
match `hashlib` behavior:
- enable `update` after `digest`
- add `copy` method
- add `digest_size`, `block_size` and `name` readonly attributes

Signed-off-by: Aurélien Lambert <aure@olli-ai.com>

...

Dobatymo commented:
Hi! Thank you very much for this PR. I agree that the extra copy is worth improved compatibility with `hashlib`.

It seems `_notfinal` is not necessary anymore. Could you remove it from the code?

Also could you improve the tests a little bit? Maybe instead of modifying the current one, add a new one, which compares the value of `digest()` before and after update.

aure-olli commented:
Hi

Before I kept `_notfinal` to avoid calling `Finalize` each time `digest` is called, but actually I don't think that's a worthy optimization (no reason to call `digest` several times), so I've also cleaned `_hash`.

While I was at it, I added `copy`, `digest_size`, `block_size` and `name` to match the behavior of `hashlib`. Especially `copy` makes lot of sense with my `update` optimization.

I did that by allowing `MetroHash128(MetroHash128())`, I don't know how else it can be done. Using `@property` for the readonly attributes added quite a lot of `.cpp` code, but I don't think it's much a problem.

Dobatymo commented:
Thanks a lot!

I don't really like how `seed` is overloaded for the `copy()`, but I will leave it until I find a better way.
